2020 JS Chapter (#1652)
HTTPArchive/almanac.httparchive.org: src/content/en/2020/javascript.md

  caption="Distribution of the amount of wasted JavaScript bytes per mobile page.",
  description="Bar chart showing the distribution of amount of wasted JavaScript bytes per page. From the 10, 25, 50, 75, and 90th percentiles, the distribution goes: 0, 57, 153, 335, and 568 KB.",
  chart_url="https://docs.google.com/spreadsheets/d/e/2PACX-1vRn1IaMxnTl0jhdC-C-vC5VLN_boJfLAaOfGJ968IalK1vPc8-dz0OkVmNY0LjMxZ6BIwSRB7xtRmIE/pubchart?oid=773002950&format=interactive",
  sheets_gid="611267860",
  sql_file="unused_js_bytes_distribution.sql"
) }}

As raw numbers, those may or may not jump out at you depending on how much of a performance nut you are, but when you look at it as a percentage of the total JavaScript used on each page, it becomes a bit easier to see just how much waste we're sending.

{{ figure_markup(
  caption="Percent of the median mobile page's JavaScript bytes that are unused.",
  content="37.22%",
  classes="big-number",
  sheets_gid="611267860",
  sql_file="unused_js_bytes_distribution.sql"
) }}

That 153 KB equates to ~37% of the total script size that we send down to mobile devices. There's definitely some room for improvement here.

### Request Count

> Another way of looking at how much JavaScript we use is to explore how many JavaScript requests are made on each page. While reducing the number of requests was paramount to maintaing good performance with HTTP 1.1, with HTTP/2 the opposite is the case: breaking JavaScript down into smaller, individual files is ideal for performance. 

Reviewer: We should reference our own HTTP/2 chapter on this before external sources if they cover this.

If I can push my own reference I touch on this in last year's chapter and it's not as clear cut as this sentence makes out: http://127.0.0.1:8080/en/2019/http2#impact-of-http2

> Another way of looking at how much JavaScript we use is to explore how many JavaScript requests are made on each page. While reducing the number of requests was paramount to maintaing good performance with HTTP 1.1, with HTTP/2 the opposite is the case: breaking JavaScript down into smaller, individual files is ideal for performance. 

Author: Sorry for not already linking up some external resources on all this stuff....I tend to fill that in after writing (going to hunt down links sometimes disrupts my flow a bit...I'm finicky. ;) )

I should add a "typically" in the sentence about breaking JS down, 'cuz yeah, there are exceptions. https://web.dev/granular-chunking-nextjs/ seemed like an interesting one to point to for recent research on this (I was never crazy about the Khan Academy post....seemed like there should be a middle point between going from 29 requests to 300 before we throw chunking out the window).

@bazzadp Your chapter does a great job of getting into the nuance. Is this year's going to as well?

> Another way of looking at how much JavaScript we use is to explore how many JavaScript requests are made on each page. While reducing the number of requests was paramount to maintaing good performance with HTTP 1.1, with HTTP/2 the opposite is the case: breaking JavaScript down into smaller, individual files is ideal for performance. 

Reviewer 2: > If I can push my own reference I touch on this in last year's chapter and it's not as clear cut as this sentence makes out: http://127.0.0.1:8080/en/2019/http2#impact-of-http2

@bazzadp your localhost is your localhost! :wink:

> Another way of looking at how much JavaScript we use is to explore how many JavaScript requests are made on each page. While reducing the number of requests was paramount to maintaing good performance with HTTP 1.1, with HTTP/2 the opposite is the case: breaking JavaScript down into smaller, individual files is ideal for performance. 

Reviewer: > (I was never crazy about the Khan Academy post....seemed like there should be a middle point between going from 29 requests to 300 before we throw chunking out the window).

100%!

> @bazzadp Your chapter does a great job of getting into the nuance. Is this year's going to as well?

No there's not so much on this particular subject in this year's HTTP/2 chapter.

> @bazzadp your localhost is your localhost! 😉

D'Oh! How does this interweb thing work again? https://almanac.httparchive.org/en/2019/http2#impact-of-http2

> Another way of looking at how much JavaScript we use is to explore how many JavaScript requests are made on each page. While reducing the number of requests was paramount to maintaing good performance with HTTP 1.1, with HTTP/2 the opposite is the case: breaking JavaScript down into smaller, individual files is ideal for performance. 

Reviewer 2: > D'Oh! How does this interweb thing work again?

I think you have way too many browser tabs/windows open (like we all do) or the morning coffee was not strong enough. ;-)

  image="bytes-by-3p-mobile.png",
  caption="Distribution of the number of JavaScript bytes by host for mobile.",
  description="Bar chart showing the distribution of JavaScript bytes per host for mobile. The 10, 25, 50, 75, and 90th percentiles for first party bytes are: 18, 60, 134, 275, and 560 KB. These values are consistently smaller than the desktop values, but only by 10-30 KB. Similar to desktop, the third party bytes are higher than first party, on mobile this difference is not as wide, only up to 328 KB at the 90th percentile.",
  chart_url="https://docs.google.com/spreadsheets/d/e/2PACX-1vRn1IaMxnTl0jhdC-C-vC5VLN_boJfLAaOfGJ968IalK1vPc8-dz0OkVmNY0LjMxZ6BIwSRB7xtRmIE/pubchart?oid=231883913&format=interactive",
  sheets_gid="1145367011",
  sql_file="bytes_by_3p.sql"
) }}

## How do we load our JavaScript?
The way we load JavaScript has a significant impact on the overall experience.

JavaScript, by default, is parser blocking. In other words, when the browser discovers a script element, it must pause parsing of the HTML until the script has been downloaded, parsed and executed. It's a significant bottleneck and a common contributor to pages that are slow to render.

> We can start to offset some of the cost of loading JavaScript by loading scripts either asynchronously (which only halts the HTML parser during the parse and execution phase) or deferred (which doesn't halt the HTML parser at all). Both attributes are only available on external scripts—inline scripts cannot have them applied.

Author: Yeah, good call.

  description="Bar chart showing the distribution of the percent of mobile JavaScript requests by compression method and host. 66% and 64% of first and third party JavaScript requests use gzip. 15% of first party and 24% of third party scripts requests use brotli. And 19% of first party and 12% of third party scripts do not have a compression method set.",
  chart_url="https://docs.google.com/spreadsheets/d/e/2PACX-1vRn1IaMxnTl0jhdC-C-vC5VLN_boJfLAaOfGJ968IalK1vPc8-dz0OkVmNY0LjMxZ6BIwSRB7xtRmIE/pubchart?oid=1402692197&format=interactive",
  sheets_gid="564760060",
  sql_file="compression_method_by_3p.sql"
) }}

Third-party scripts are also less likely to be served without any compression at all: 12% of third-party scripts have neither GZip nor Brotli applied, compared to 19% of first-party scripts.

It's worth taking a closer look those scripts that _don't_ have compression applied. Compression becomes more efficient in terms of savings the more content it has to work with. In other words, if the file is tiny, sometimes the cost of compressing the file doesn't outweight the miniscule reduction in file size.

> Thankfully, that's exactly what we see, particularly in third-party scripts where 90% of uncompressed scripts are less than 5kb in size. On the other hand, 49% of uncompressed first-party scripts are less than 5kb and 37% of uncompressed first-party scripts are over 10kb. So while we do see a lot of small uncompressed first-party scripts, there are still quite a few that would benefit from some compression.

Author: :)

  chart_url="https://docs.google.com/spreadsheets/d/e/2PACX-1vRn1IaMxnTl0jhdC-C-vC5VLN_boJfLAaOfGJ968IalK1vPc8-dz0OkVmNY0LjMxZ6BIwSRB7xtRmIE/pubchart?oid=649523941&format=interactive",
  sheets_gid="2035770638",
  sql_file="correlations.sql"
) }}

The above chart uses the Pearson coefficient of correlation. There's a long, kinda complex definition of what that means precisely, but the gist is that we're looking for the strength of the correlation between two different numbers. If we find a coefficient of 1, we'd have a direct correlation. A correlation of 0 would show no connection between two numbers. Anything below zero shows a negative correlation—in other words, as one number goes up the other one decreases.

> First, there doesn't seem to be much measurable correlation between our JavaScript metrics and the Lighthouse Accessibility score here. That stands in stark opposition to what's been found elsewhere, notably through WebAim's annual research.

Reviewer 2: > ... notably through WebAim's annual research

Linking to the report, if possible, would be great.

The most likely explanation for this is that Lighthouse's accessibility tests aren't as comprehensive (yet!) as what is available through other tools, like WebAIM, that have accessibility as their primary focus.

Where we do see a strong correlation is between the amount of JavaScript bytes and both the overall Lighthouse performance score and Total Blocking Time.

> The correlation between JavaScript bytes and Lighthouse performance scores is -.47. In other words, as JS bytes increase, Lighthouse performance scores decrease. The overall bytes has a stronger correlation than the amount of third-party scripts, hinting that while they certainly play a role, we can't place all the blame on third-parties.

Reviewer 2: Writing fractions without the leading `0` is difficult to read in plain text and the dot is easy to miss.

```suggestion
The correlation between JavaScript bytes and Lighthouse performance scores is -0.47. In other words, as JS bytes increase, Lighthouse performance scores decrease. The overall bytes has a stronger correlation than the amount of third-party scripts, hinting that while they certainly play a role, we can't place all the blame on third-parties.
```

> The connection between Total Blocking Time and JavaScript bytes is even more signficant (.55 for overall bytes, .48 for third-party bytes). That's not too surprising given what we know about all the work browsers have to do to get JavaScript to run in a page—more bytes means more time.

Reviewer 2: ```suggestion
The connection between Total Blocking Time and JavaScript bytes is even more significant (0.55 for overall bytes, 0.48 for third-party bytes). That's not too surprising given what we know about all the work browsers have to do to get JavaScript to run in a page—more bytes means more time.
```

### Security Vulnerabilities

> One other helpful audit that Lighthouse runs is to check for known security vulnerabilities in third-party libraries. It does this by detecting which libraries and frameworks are used on a given page, and what version is used of each. Then it checks Snyk's open-source vulnerability database to see what vulnerabilities have been discovered in the identified tools.

Reviewer 2: > Snyk's open-source vulnerability database

Link?

It's important that we think critically about the tools we use to build the web and what the trade-offs are, so it makes sense to look closely at what we see in use today.

### Libraries
HTTP Archive uses Wappalyzer to detect technologies in use on a given page. Wappalazyer tracks both JavaScript Libraries (think of these as a collection of snippets or helper functions to ease development, like jQuery) and JavaScript Frameworks (these are more likely scaffolding and provide templating and structure, like React).

The popular libraries in use are largely unchanged from last year, with jQuery continuing to dominate usage and only one of the top 21 libraries falling out (lazy.js, replaced by DataTables). In fact, even the percentages of the top libraries has barely changed from last year. 

> {{ table? showing rank, library, percentage and last years rank}}

Reviewer: Change to Jinja2 comment. This will allown the build to work, so give you actual errors in the GitHub tests.

```suggestion
{# table? showing rank, library, percentage and last years rank #}
```

  caption="Distribution of the amount of wasted JavaScript bytes per mobile page.",
  description="Bar chart showing the distribution of amount of wasted JavaScript bytes per page. From the 10, 25, 50, 75, and 90th percentiles, the distribution goes: 0, 57, 153, 335, and 568 KB.",
  chart_url="https://docs.google.com/spreadsheets/d/e/2PACX-1vRn1IaMxnTl0jhdC-C-vC5VLN_boJfLAaOfGJ968IalK1vPc8-dz0OkVmNY0LjMxZ6BIwSRB7xtRmIE/pubchart?oid=773002950&format=interactive",
  sheets_gid="611267860",
  sql_file="unused_js_bytes_distribution.sql"
) }}

As raw numbers, those may or may not jump out at you depending on how much of a performance nut you are, but when you look at it as a percentage of the total JavaScript used on each page, it becomes a bit easier to see just how much waste we're sending.

{{ figure_markup(
  caption="Percent of the median mobile page's JavaScript bytes that are unused.",
  content="37.22%",
  classes="big-number",
  sheets_gid="611267860",
  sql_file="unused_js_bytes_distribution.sql"
) }}

That 153 KB equates to ~37% of the total script size that we send down to mobile devices. There's definitely some room for improvement here.

### Request Count

> Another way of looking at how much JavaScript we use is to explore how many JavaScript requests are made on each page. While reducing the number of requests was paramount to maintainingg good performance with HTTP/1.1, with HTTP/2 the opposite is the case: breaking JavaScript down into [smaller, individual files](https://web.dev/granular-chunking-nextjs/) is [typically better for performance](https://almanac.httparchive.org/en/2019/http2#impact-of-http2). 

Reviewer 3: ```suggestion
Another way of looking at how much JavaScript we use is to explore how many JavaScript requests are made on each page. While reducing the number of requests was paramount to maintaining good performance with HTTP/1.1, with HTTP/2 the opposite is the case: breaking JavaScript down into [smaller, individual files](https://web.dev/granular-chunking-nextjs/) is [typically better for performance](https://almanac.httparchive.org/en/2019/http2#impact-of-http2). 
```

      <tr>
        <td>Vue.js</td>
        <td class="numeric">0.2%</td>
        <td>&nbsp;</td>
      </tr>
      <tr>
        <td>AngularJS, jQuery</td>
        <td>&nbsp;</td>
        <td class="numeric">0.2%</td>
      </tr>
      <tr>
        <td>GSAP, Hammer.js, Lodash, React, RequireJS, Zepto</td>
        <td class="numeric">0.2%</td>
        <td>&nbsp;</td>
      </tr>
    </tbody>
    <tfoot>
      <tr>
        <th>Grand Total</th>
        <th class="numeric">1.7%</th>
        <th class="numeric">1.4%</th>
      </tr>
    </tfoot>
  </table>
  <figcaption>
    {{ figure_link(
      caption="The most popular combinations of React, Angular, and Vue with and without jQuery.",
      sheets_gid="795160444",
      sql_file="frameworks_libraries_combos.sql"
    ) }}
  </figcaption>
</figure>

More importantly, all these tools typically mean more code and more processing time.

Looking specfically at the frameworks in use, we see that the median JavaScript bytes for pages using them varies dramatically depending on _what_ is being used.

The graph below shows the median bytes for pages where any of the top 35 most commonly detected frameworks were found, broken down by client.

> {# Median bytes by JS framework #}

Reviewer 3: @tkadlec do we have this data already?

Author: Yup! https://docs.google.com/spreadsheets/d/1cgXJrFH02SHPKDGD0AelaXAdB3UI7PIb5dlS0dxVtfY/edit#gid=1206934500

Reviewer 3: Oh perfect, I didn't see the new queries in this PR 👍

In fact, the dominance of jQuery is supported even further when you stop to consider that, of the top 10 libraries, 6 of them are either jQuery or require jQuery in order to be used (jQuery UI, jQuery Migrate, FancyBox, Lightbox and Slick).

### Frameworks

> When we look at the frameworks, we also don't see much of a dramatic change in terms of adoption in the main frameworks that were highlighted last year. Vue.js has seen a significant increase, and AMP grew a bit, but most of them are more less where they were a year ago. 

Reviewer 2: ```suggestion
When we look at the frameworks, we also don't see much of a dramatic change in terms of adoption in the main frameworks that were highlighted last year. Vue.js has seen a significant increase, and AMP grew a bit, but most of them are more or less where they were a year ago. 
```

{# Compare same frameworks from last year's chapter to this year in bar chart? #}

> It's worth noting that the detection issue that was noted last year still applies, and still impacts th results here. It's possible that there _has_ been a significant change in popularity for a few more of these tools, but we just don't see it with the way the data is currently collected.

Reviewer 2: ```suggestion
It's worth noting that the detection issue that was noted last year still applies, and still impacts the results here. It's possible that there _has_ been a significant change in popularity for a few more of these tools, but we just don't see it with the way the data is currently collected.
```

{# Distribution of number of JS frameworks or libraries on a page #}

When we look at the common combinations that we see in production, most of them are to be expected. Knowing jQuery's dominance, it's unsurprising that most of the popular combinations include jQuery and any number of jQuery related plugins.

{# Top 20? Combos with percentage #}

We do also see a fair amount of more "modern" frameworks, like React, Vue and Angular—paired with jQuery, like as a result of either migration or third-parties including one or the other.

{# React/Angular/Vue % with and without jQuery? #}

More importantly, all these tools typically mean more code and more processing time.

Looking specfically at the frameworks in use, we see that the median JavaScript bytes for pages using them varies dramatically depending on _what_ is being used.

The graph below shows the median bytes for pages where any of the top 35 most commonly detected frameworks were found, broken down by client.

{# Median bytes by JS framework #}

> On one of the spectrum are frameworks like React or Angular or Ember, which tend to ship a lot of code regardless of the client. On the other end, we see minimalist frameworks like Alpine.js and Svelte showing very promising results. Defaults are very important, and it seems that by starting with highly performant defaults, Svelte and Alpine are both succeeding (so far...the sample size is pretty small) in creating a lighter set of pages.

Reviewer 2: ```suggestion
On one of the spectrum are frameworks like React, Angular, and Ember, which tend to ship a lot of code regardless of the client. On the other end, we see minimalist frameworks like Alpine.js and Svelte showing very promising results. Defaults are very important, and it seems that by starting with highly performant defaults, Svelte and Alpine are both succeeding (so far... the sample size is pretty small) in creating a lighter set of pages.
```

> We get a very similar picture when looking at main thread time for pages where these tools were detected.

Reviewer 2: ```suggestion
We get a very similar picture when looking at main thread execution time for pages where these tools were detected.
```

{# Median main thread by JS framework #}

> Ember's mobile main thread time jumps out and kind of distorts the graph with how long it takes. Pulling it out makes the picture a bit easier to understand. 

Reviewer 2: ```suggestion
Ember's mobile main thread execution time jumps out and kind of distorts the graph with how long it takes. Pulling it out makes the picture a bit easier to understand. 
```

Author: I'm gonna leave that out here as it's not just execution time being recorded, but any main thread time (parse/compile/execution).

{# Median main thread by JS framework, no Ember #}

> Tools like React, GASP and RequireJS tend to spend a lot of time on the main thread of the browser, regardless of whether it's a desktop or mobile page view. The same tools that tend to lead to less code overall—tools like Alpine and Svelte—also tend to lead to lower impact on the main thread.

Reviewer 2: I forgot if we are using Oxford comma or not, but we should be consistent.

```suggestion
Tools like React, GASP, and RequireJS tend to spend a lot of time on the main thread of the browser, regardless of whether it's a desktop or mobile page view. The same tools that tend to lead to less code overall—tools like Alpine and Svelte—also tend to lead to lower impact on the main thread.
```
